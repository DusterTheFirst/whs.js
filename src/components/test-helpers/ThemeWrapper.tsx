/*!
 * Copyright (C) 2018-2020  Zachary Kohnen (DusterTheFirst)
 */

import React from "react";
import { ThemeProvider } from "styled-components/native";
import { lightTheme } from "../../styles/theme";

/** The wrapper around a tested component to provide it with a theme */
const ThemeWrapper: React.FC = ({ children }) => (
    <ThemeProvider theme={lightTheme}>{children}</ThemeProvider>
);

export default ThemeWrapper;