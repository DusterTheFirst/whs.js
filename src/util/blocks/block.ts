/*!
 * Copyright (C) 2018-2019  Zachary Kohnen (DusterTheFirst)
 */

export enum Block {
    A = "a",
    B = "b",
    C = "c",
    D = "d",
    E = "e",
    F = "f"
}