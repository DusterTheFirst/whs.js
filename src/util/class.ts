/*!
 * Copyright (C) 2018-2019  Zachary Kohnen (DusterTheFirst)
 */

import { Dayjs } from "dayjs";
import { BlockColor } from "./blocks/blockColor";

export default interface IClass {
    teacher: string;
    room: number;
    name: string;
    block: BlockColor;
    meets: number;
}

export interface ITimes {
    start: Dayjs;
    end: Dayjs;
}