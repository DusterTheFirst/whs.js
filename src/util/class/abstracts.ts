/*!
 * Copyright (C) 2018-2019  Zachary Kohnen (DusterTheFirst)
 */

import { IDR, IElective, IFree, IMajor } from "./blocks";
import { IClassMeta, ITimes } from "./extentions";

/** A type that encompasses any type of class there could be */
export type Class = IElective & IMajor & IDR;
/** A class with all of the information needed to process it */
export type StorageClass = Class & IClassMeta;
export type DisplayClass = (Class | IFree) & ITimes;