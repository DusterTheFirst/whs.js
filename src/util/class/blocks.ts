/*!
 * Copyright (C) 2018-2019  Zachary Kohnen (DusterTheFirst)
 */

import { BlockColor } from "../blocks/blockColor";
import { IAdvisedClass, IIrregular } from "./primitives";
import { ClassType, IClassType } from "./type";

/** The most basic information needed to identify a major */
export interface IMajor extends IAdvisedClass, IClassType<ClassType.Major> {
    /** The name of the major */
    name: string;
    /** The block color which the major meets */
    block: BlockColor;
    /** If the major has a Lab block or not */
    lab: boolean;
}

/** The most basic information needed to identify an elective */
export interface IElective extends IAdvisedClass, IIrregular, IClassType<ClassType.Elective> {
    /** The name of the elective */
    name: string;
    /** The block color which the elective meets (None = 730) */
    block: BlockColor;
}

/** The most basic information needed to identify a DR */
export interface IDR extends IAdvisedClass, IIrregular, IClassType<ClassType.DR> {
    /** The block color which the elective meets */
    block: BlockColor;
}

/** The most basic information needed to identify a Free block */
export interface IFree extends IIrregular, IClassType<ClassType.Free> {
    /** The block color which the elective meets */
    block: BlockColor;
}