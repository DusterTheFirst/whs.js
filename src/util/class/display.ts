/*!
 * Copyright (C) 2018-2019  Zachary Kohnen (DusterTheFirst)
 */

import { BlockColor } from "../blocks/blockColor";
import { ITimes } from "./extentions";
import { IAdvisedClass } from "./primitives";

/** The display information for a class block */
export interface IClassBlock extends IAdvisedClass, ITimes {
    /** The name of the class block */
    name: string;
    /** The block color which to display for the meets */
    block: BlockColor;
}

/** The most basic information needed to identify a Free block */
export interface IFreeBlock extends ITimes {
    /** The block color which the class meets */
    block: BlockColor;
}