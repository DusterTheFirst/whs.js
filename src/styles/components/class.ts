/*!
 * Copyright (C) 2018-2020  Zachary Kohnen (DusterTheFirst)
 */

import { BlockColor } from "@whsha/classes/v2/block";
import styled from "styled-components/native";
import { getDisplayColorForBlock } from "../blockColor";
import { DimText } from "./common";

/** Text to be placed on the left of the class display */
export const LeftClassText = styled(DimText)({
    flex: 1,
    textAlign: "left",
});

/** Text to be placed on the right of the class display */
export const RightClassText = styled(DimText)({
    flex: 1,
    textAlign: "right",
});

/** Text to be placed for display as times */
export const TimesText = styled(DimText)({
    paddingLeft: 5,
    textAlign: "right"
});

/** Properties for a styled component with a class color prop */
interface IClassColoredComponentProps {
    /** The class color to show */
    classColor?: BlockColor;
}

/** The text to show the colorblind labels with */
export const ColorblindLabelText = styled.Text<IClassColoredComponentProps>(
    (props) => ({
        color:
            getDisplayColorForBlock(props.classColor)
            ?? props.theme.colors.dimText,
        fontWeight: "bold",
        paddingLeft: 10,
        paddingRight: 10,
        textAlign: "center"
    })
);

/** The text to show the title of a class */
export const ClassTitleText = styled(LeftClassText)<IClassColoredComponentProps>(
    (props) => ({
        color:
            getDisplayColorForBlock(props.classColor)
            ?? props.theme.colors.title,
        fontSize: 25,
        fontWeight: "bold",
    })
);

/** A view for placing an item in the center of a class view */
export const MiddleClassView = styled.View({
    alignItems: "center",
    flex: 1,
    justifyContent: "center",
    maxWidth: "75px"
});

/** A class view row */
export const ClassViewRow = styled.View({
    flexDirection: "row",
    flexWrap: "nowrap",
    justifyContent: "space-between"
});

/** The info row for the class view */
export const ClassViewInfoRow = styled(ClassViewRow)({
    marginTop: 5
});

/** Props for a styled component that can have a dynamic flex state */
interface IFlexableProps {
    /** The flex state */
    doFlex: boolean;
}

/** The view to surround the times part of the class */
export const TimesView = styled.View<IFlexableProps>(props => ({
    alignItems: "center",
    flex: props.doFlex ? 1 : undefined,
    flexDirection: "row",
    justifyContent: "flex-end"
}));

/** The view for containing a class */
export const ClassContainerView = styled.View(props => ({
    alignItems: "center",
    backgroundColor: props.theme.colors.classBackground,
    justifyContent: "center",
    padding: 10,
    width: "100%"
}));