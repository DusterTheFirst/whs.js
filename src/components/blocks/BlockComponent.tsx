/*!
 * Copyright (C) 2018-2020  Zachary Kohnen (DusterTheFirst)
 */

import React, { memo } from "react";
import { IDR, IMajor, IMinor } from "../../util/class/classes";
import { ITimes } from "../../util/class/extentions";
import { IColored } from "../../util/class/primitives";
import { isMajor, isMinor } from "../../util/class/type";
import ClassComponent from "./ClassComponent";
import FreeComponent from "./FreeComponent";

/** The parameters to pass to a BlockComponent */
interface IDisplayBlock extends IColored, ITimes {
    /** The class to display */
    clazz?: IMajor | IMinor | IDR;
}

/** A component that will display a block of many variety */
function BlockComponent({ end, start, clazz, block }: IDisplayBlock) {
    if (clazz === undefined) {
        return <FreeComponent start={start} end={end} block={block} />;
    } else {
        if (isMajor(clazz) || isMinor(clazz)) {
            return (
                <ClassComponent
                    start={start}
                    end={end}
                    name={clazz.name}
                    block={clazz.block}
                    room={clazz.room}
                    teacher={clazz.teacher}
                />
            );
        } else {
            return (
                <ClassComponent
                    start={start}
                    end={end}
                    name="Directed Reaserch"
                    block={clazz.block}
                    room={clazz.room}
                    teacher={clazz.teacher}
                />
            );
        }
    }
}

export default memo(BlockComponent);