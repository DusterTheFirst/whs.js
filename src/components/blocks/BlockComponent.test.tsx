/*!
 * Copyright (C) 2018-2020  Zachary Kohnen (DusterTheFirst)
 */

import { BlockColor } from "@whsha/classes/v2/block";
import { IClass } from "@whsha/classes/v2/class";
import { SchoolDay } from "@whsha/classes/v2/schoolDay";
import { Semester } from "@whsha/classes/v2/semester";
import dayjs from "dayjs";
import { default as useCustomFormat } from "dayjs/plugin/customParseFormat";
import React from "react";
import { render } from "react-native-testing-library";
import ThemeWrapper from "../test-helpers/ThemeWrapper";
import BlockComponent from "./BlockComponent";

dayjs.extend(useCustomFormat);

/** A dummy class to use for tests */
export const dummyClass: IClass = {
    block: BlockColor.Red,
    lab: false,
    lunches: {},
    meets: {
        [SchoolDay.One]: false,
        [SchoolDay.Two]: false,
        [SchoolDay.Three]: false,
        [SchoolDay.Four]: false,
        [SchoolDay.Five]: false,
        [SchoolDay.Six]: false,
        [SchoolDay.Seven]: false
    },
    name: "",
    room: "",
    semesters: {
        [Semester.First]: true,
        [Semester.Second]: true
    },
    teacher: "",
    uuid: ""
};

it("Renders BlockComponent with a class", () => {
    const comp = render(
        (
            <BlockComponent clazz={dummyClass} block={BlockColor.Red} start={dayjs("7:30 AM", "h:mm A")} end={dayjs("8:30 AM", "h:mm A")} />
        ),
        { wrapper: ThemeWrapper }
    );

    expect(comp.toJSON()).toMatchSnapshot();
});

it("Renders BlockComponent with a free", () => {
    const comp = render(
        (
            <BlockComponent block={BlockColor.None} start={dayjs("7:30 AM", "h:mm A")} end={dayjs("8:30 AM", "h:mm A")} />
        ),
        { wrapper: ThemeWrapper }
    );

    expect(comp.toJSON()).toMatchSnapshot();
});