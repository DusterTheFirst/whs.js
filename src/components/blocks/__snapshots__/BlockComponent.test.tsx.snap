// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders a free 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#ffffff",
        "justifyContent": "center",
        "paddingBottom": 10,
        "paddingLeft": 10,
        "paddingRight": 10,
        "paddingTop": 10,
        "width": "100%",
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
          "flexWrap": "nowrap",
          "justifyContent": "space-between",
        },
      ]
    }
  >
    <Text
      classColor="No Color"
      ellipsizeMode="middle"
      numberOfLines={1}
      style={
        Array [
          Object {
            "color": "#999999",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "fontSize": 25,
            "fontWeight": "bold",
            "textAlign": "left",
          },
        ]
      }
    >
      Free
    </Text>
    <View
      doFlex={false}
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "flex-end",
          },
        ]
      }
    >
      <Text
        numberOfLines={1}
        style={
          Array [
            Object {
              "color": "#808080",
              "paddingLeft": 5,
              "textAlign": "right",
            },
          ]
        }
      >
        7:30
         - 
        8:30 AM
      </Text>
    </View>
  </View>
</View>
`;

exports[`Tests different variations of classes Renders BlockComponent with a dr 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#ffffff",
        "justifyContent": "center",
        "paddingBottom": 10,
        "paddingLeft": 10,
        "paddingRight": 10,
        "paddingTop": 10,
        "width": "100%",
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
          "flexWrap": "nowrap",
          "justifyContent": "space-between",
        },
      ]
    }
  >
    <Text
      classColor="Green"
      ellipsizeMode="middle"
      numberOfLines={1}
      style={
        Array [
          Object {
            "color": "#5da539",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "fontSize": 25,
            "fontWeight": "bold",
            "textAlign": "left",
          },
        ]
      }
    >
      Directed research
    </Text>
    <View
      doFlex={false}
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "flex-end",
          },
        ]
      }
    >
      <Text
        numberOfLines={1}
        style={
          Array [
            Object {
              "color": "#808080",
              "paddingLeft": 5,
              "textAlign": "right",
            },
          ]
        }
      >
        12:14
         - 
        1:24 PM
      </Text>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
          "flexWrap": "nowrap",
          "justifyContent": "space-between",
          "marginTop": 5,
        },
      ]
    }
  >
    <Text
      numberOfLines={1}
      style={
        Array [
          Object {
            "color": "#808080",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "textAlign": "left",
          },
        ]
      }
    >
      Dr. Test
    </Text>
    <Text
      numberOfLines={1}
      style={
        Array [
          Object {
            "color": "#808080",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "textAlign": "right",
          },
        ]
      }
    >
      Room 102
    </Text>
  </View>
</View>
`;

exports[`Tests different variations of classes Renders BlockComponent with a major 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#ffffff",
        "justifyContent": "center",
        "paddingBottom": 10,
        "paddingLeft": 10,
        "paddingRight": 10,
        "paddingTop": 10,
        "width": "100%",
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
          "flexWrap": "nowrap",
          "justifyContent": "space-between",
        },
      ]
    }
  >
    <Text
      classColor="Red"
      ellipsizeMode="middle"
      numberOfLines={1}
      style={
        Array [
          Object {
            "color": "#c93a3a",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "fontSize": 25,
            "fontWeight": "bold",
            "textAlign": "left",
          },
        ]
      }
    >
      Test Major
    </Text>
    <View
      doFlex={false}
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "flex-end",
          },
        ]
      }
    >
      <Text
        numberOfLines={1}
        style={
          Array [
            Object {
              "color": "#808080",
              "paddingLeft": 5,
              "textAlign": "right",
            },
          ]
        }
      >
        7:30
         - 
        8:30 AM
      </Text>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
          "flexWrap": "nowrap",
          "justifyContent": "space-between",
          "marginTop": 5,
        },
      ]
    }
  >
    <Text
      numberOfLines={1}
      style={
        Array [
          Object {
            "color": "#808080",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "textAlign": "left",
          },
        ]
      }
    >
      Mr. Test
    </Text>
    <Text
      numberOfLines={1}
      style={
        Array [
          Object {
            "color": "#808080",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "textAlign": "right",
          },
        ]
      }
    >
      Fitness center
    </Text>
  </View>
</View>
`;

exports[`Tests different variations of classes Renders BlockComponent with a minor 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#ffffff",
        "justifyContent": "center",
        "paddingBottom": 10,
        "paddingLeft": 10,
        "paddingRight": 10,
        "paddingTop": 10,
        "width": "100%",
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
          "flexWrap": "nowrap",
          "justifyContent": "space-between",
        },
      ]
    }
  >
    <Text
      classColor="Yellow"
      ellipsizeMode="middle"
      numberOfLines={1}
      style={
        Array [
          Object {
            "color": "#e9dc0f",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "fontSize": 25,
            "fontWeight": "bold",
            "textAlign": "left",
          },
        ]
      }
    >
      Test minor
    </Text>
    <View
      doFlex={false}
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "flex-end",
          },
        ]
      }
    >
      <Text
        numberOfLines={1}
        style={
          Array [
            Object {
              "color": "#808080",
              "paddingLeft": 5,
              "textAlign": "right",
            },
          ]
        }
      >
        12:14
         - 
        1:24 PM
      </Text>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "flexDirection": "row",
          "flexWrap": "nowrap",
          "justifyContent": "space-between",
          "marginTop": 5,
        },
      ]
    }
  >
    <Text
      numberOfLines={1}
      style={
        Array [
          Object {
            "color": "#808080",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "textAlign": "left",
          },
        ]
      }
    >
      Mrs. Test
    </Text>
    <Text
      numberOfLines={1}
      style={
        Array [
          Object {
            "color": "#808080",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "textAlign": "right",
          },
        ]
      }
    >
      Room 334
    </Text>
  </View>
</View>
`;
